Add Dynamic HalfOpenPeriod and WhileHalfOpen ResetMode for Improved Retry Handling

gian-hancock opened the pull request:
# Context
I'm opening this PR to see if you're interested in upstreaming some enhancements from my fork, designed to better handle specific retry scenarios.

I needed to address the following challenges:
- Messages are sent to a DLQ after too many retries (this behavior is outside our control).
- When an upstream service fails for extended periods, we want to prevent excessive retries while still allowing recovery.
- Initial retries shouldn’t be excessively delayed to accommodate transient failures.

To solve this, I've implemented a dynamic HalfOpenPeriod that enables exponential backoff. After an error closes the circuit, it initially enters HalfOpen quickly, but each subsequent attempt takes progressively longer if failures persist. Additionally, I introduced a new ResetMode to complement this behavior.

# Feature Description

## Dynamic HalfOpenPeriod:
HalfOpenPeriod can now vary based on how many consecutive times the circuit enters the HalfOpen state. This enables exponential backoff.

```c#
                o.EnableCircuitBreaker(
                    c => c.OpenOn<MyCustomException>(
                        attempts: 1, 
                        trackingPeriodInSeconds: 60,
                        // halfOpenPeriod = 30s, 90s, 270s, etc...
                        halfOpenPeriodProvider: attemptCount => TimeSpan.FromSeconds(30 * Math.Pow(3, attemptCount)),                      
                        resetIntervalInSeconds: 60,
                        resetMode: ResetMode.WhileHalfOpen));
```

## WhileHalfOpen ResetMode:
Introduces two ResetModes:
- **WhileAnyState (default)**: Matches the current behavior.
- **WhileHalfOpen**: The circuit resets (closes) only while in HalfOpen state, with the reset interval starting when it enters HalfOpen.

### Motivation:

Prevents resets from occurring while the circuit is still open when ResetInterval < HalfOpenPeriod. This is crucial for dynamic HalfOpenPeriods, ensuring that ResetInterval remains meaningful across varying HalfOpenPeriods.

# Summary
I've taken care to implement the changes in a backwards compatible way. Let me know if you'd like to incorporate this into the project, and if any modifications or further explanations are needed.

---
Rebus is [MIT-licensed](https://opensource.org/licenses/MIT). The code submitted in this pull request needs to carry the MIT license too. By leaving this text in, __I hereby acknowledge that the code submitted in the pull request has the MIT license and can be merged with the Rebus codebase__.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/rebus-org/Rebus.CircuitBreaker?pullRequest=11) <br/>All committers have signed the CLA.

gian-hancock committed:
Implement Dynamic HalfOpenPeriod and WhileHalfOpen ResetMode

Dynamic HalfOpenPeriod:

HalfOpenPeriod can now vary based on how many consecutive times the circuit enters the HalfOpen state.
This enables features like exponential backoff.

WhileHalfOpen ResetMode:

Introduces two ResetModes:
- WhileAnyState (default): Matches the current behavior.
- WhileHalfOpen: The circuit resets (closes) only while in HalfOpen state, with the reset interval starting when it enters HalfOpen.

Motivation:

Prevents resets from occurring while the circuit is still open when ResetInterval < HalfOpenPeriod.
This is crucial for dynamic HalfOpenPeriods, ensuring that ResetInterval remains meaningful across varying HalfOpenPeriods.

gian-hancock commented:
I'll investigate why the CI is failing and come back with an update.

gian-hancock committed:
Adjust timing thresholds in dynamic HalfOpenPeriod test

The WaitDynamicHalfOpenPeriodBeforeHalfOpening test was flaky due to
overly tight timing constraints. Since the CircuitBreaker updates every
two seconds, these thresholds occasionally led to inconsistencies. This
change adjusts the timings to improve test stability.

gian-hancock commented:
> I'll investigate why the CI is failing and come back with an update.

I found and fixed an issue with one of the new tests. See https://github.com/rebus-org/Rebus.CircuitBreaker/pull/11/commits/f4846d0e4013abb9b7fc520aee8202bcb590a4d5.
